File: src/vb/Samples/Samples/Samples/SolidEdge/Part/PlaceFeatureLibrary.vb
Imports ApiSamples.Samples.SolidEdge
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text

Namespace ApiSamples.Samples.SolidEdge.Part
	''' <summary>
	''' Creates a new part and places features from an existing document.
	''' </summary>
	Friend Class PlaceFeatureLibrary
		Private Shared Sub RunSample(ByVal breakOnStart As Boolean)
			If breakOnStart Then
				System.Diagnostics.Debugger.Break()
			End If

			Dim application As SolidEdgeFramework.Application = Nothing
			Dim documents As SolidEdgeFramework.Documents = Nothing
			Dim partDocument As SolidEdgePart.PartDocument = Nothing
			Dim refPlanes As SolidEdgePart.RefPlanes = Nothing
			Dim refPlane As SolidEdgePart.RefPlane = Nothing
			Dim selectSet As SolidEdgeFramework.SelectSet = Nothing
			Dim features As Array = Array.CreateInstance(GetType(Object), 0)
			Dim extrudedProtrustion As SolidEdgePart.ExtrudedProtrusion = Nothing
			Dim round As SolidEdgePart.Round = Nothing

			Try
				' Register with OLE to handle concurrency issues on the current thread.
				OleMessageFilter.Register()

				' Connect to or start Solid Edge.
				application = ApplicationHelper.Connect(True, True)

				' Get a reference to the documents collection.
				documents = application.Documents

				' Create a new part document.
				partDocument = documents.AddPartDocument()

				' Always a good idea to give SE a chance to breathe.
				application.DoIdle()

				' Get a reference to the RefPlanes collection.
				refPlanes = partDocument.RefPlanes

				' Get a reference to the top RefPlane using extension method.
				refPlane = refPlanes.GetTopPlane()

				' Get path to Solid Edge training directory.  Typically, 'C:\Program Files\Solid Edge XXX\Training'.
				Dim trainingDirectory As New DirectoryInfo(InstallDataHelper.GetTrainingFolderPath())

				' Build path to source part document.
				Dim LibName As String = Path.Combine(trainingDirectory.FullName, "block.par")

				' This method will take all features from block.par and place them into the new part document.
				partDocument.PlaceFeatureLibrary(LibName, refPlane, 0.0, 0.0, 0.0, features)

				' Optionally, iterate through all of the added features.
				For Each feature As Object In features
					' Use ReflectionHelper class to get the feature type.
					Dim featureType As SolidEdgePart.FeatureTypeConstants = ReflectionHelper.GetPartFeatureType(feature)

					' Depending on the feature type, we can cast the weakly typed feature to a strongly typed feature.
					Select Case featureType
						Case SolidEdgePart.FeatureTypeConstants.igExtrudedProtrusionFeatureObject
							extrudedProtrustion = DirectCast(feature, SolidEdgePart.ExtrudedProtrusion)
						Case SolidEdgePart.FeatureTypeConstants.igRoundFeatureObject
							round = DirectCast(feature, SolidEdgePart.Round)
					End Select
				Next feature

				' Get a reference to the ActiveSelectSet.
				selectSet = application.ActiveSelectSet

				' Empty ActiveSelectSet.
				selectSet.RemoveAll()

				' Add all features to ActiveSelectSet.
				For Each feature As Object In features
					selectSet.Add(feature)
				Next feature

				' Switch to ISO view.
				application.StartCommand(SolidEdgeConstants.PartCommandConstants.PartViewISOView)
			Catch ex As System.Exception
				Console.WriteLine(ex.Message)
			Finally
				OleMessageFilter.Unregister()
			End Try
		End Sub
	End Class
End Namespace

File: src/cs/Samples/ApiSamples/Samples/Part/PlaceFeatureLibrary.cs
using SolidEdgeFramework.Extensions; //SolidEdge.Community.dll
using SolidEdgePart.Extensions; // SolidEdge.Community.dll
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace ApiSamples.Part
{
    /// <summary>
    /// Creates a new part and places features from an existing document.
    /// </summary>
    class PlaceFeatureLibrary
    {
        static void RunSample(bool breakOnStart)
        {
            if (breakOnStart) System.Diagnostics.Debugger.Break();

            SolidEdgeFramework.Application application = null;
            SolidEdgeFramework.Documents documents = null;
            SolidEdgePart.PartDocument partDocument = null;
            SolidEdgePart.RefPlanes refPlanes = null;
            SolidEdgePart.RefPlane refPlane = null;
            SolidEdgeFramework.SelectSet selectSet = null;
            Array features = Array.CreateInstance(typeof(object), 0);
            SolidEdgePart.ExtrudedProtrusion extrudedProtrustion = null;
            SolidEdgePart.Round round = null;

            try
            {
                // Register with OLE to handle concurrency issues on the current thread.
                SolidEdgeCommunity.OleMessageFilter.Register();

                // Connect to or start Solid Edge.
                application = SolidEdgeCommunity.SolidEdgeInstall.Connect(true, true);

                // Get a reference to the documents collection.
                documents = application.Documents;

                // Create a new part document.
                partDocument = documents.AddPartDocument();

                // Always a good idea to give SE a chance to breathe.
                application.DoIdle();

                // Get a reference to the RefPlanes collection.
                refPlanes = partDocument.RefPlanes;

                // Get a reference to the top RefPlane using extension method.
                refPlane = refPlanes.GetTopPlane();

                // Get path to Solid Edge training directory.  Typically, 'C:\Program Files\Solid Edge XXX\Training'.
                DirectoryInfo trainingDirectory = new DirectoryInfo(SolidEdgeCommunity.SolidEdgeInstall.GetTrainingFolderPath());

                // Build path to source part document.
                string LibName = Path.Combine(trainingDirectory.FullName, "block.par");

                // This method will take all features from block.par and place them into the new part document.
                partDocument.PlaceFeatureLibrary(LibName, refPlane, 0.0, 0.0, 0.0, out features);

                // Optionally, iterate through all of the added features.
                foreach (object feature in features)
                {
                    // Use ReflectionHelper class to get the feature type.
                    SolidEdgePart.FeatureTypeConstants featureType = ReflectionHelper.GetPartFeatureType(feature);

                    // Depending on the feature type, we can cast the weakly typed feature to a strongly typed feature.
                    switch (featureType)
                    {
                        case SolidEdgePart.FeatureTypeConstants.igExtrudedProtrusionFeatureObject:
                            extrudedProtrustion = (SolidEdgePart.ExtrudedProtrusion)feature;
                            break;
                        case SolidEdgePart.FeatureTypeConstants.igRoundFeatureObject:
                            round = (SolidEdgePart.Round)feature;
                            break;
                    }
                }

                // Get a reference to the ActiveSelectSet.
                selectSet = application.ActiveSelectSet;

                // Empty ActiveSelectSet.
                selectSet.RemoveAll();

                // Add all features to ActiveSelectSet.
                foreach (object feature in features)
                {
                    selectSet.Add(feature);
                }

                // Switch to ISO view.
                application.StartCommand(SolidEdgeConstants.PartCommandConstants.PartViewISOView);
            }
            catch (System.Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            finally
            {
                SolidEdgeCommunity.OleMessageFilter.Unregister();
            }
        }
    }
}
